apps/glusterfs/healthz.go

```diff
@@ -0,0 +1,130 @@
+//
+// Copyright (c) 2019 The heketi Authors
+//
+// This file is licensed to you under your choice of the GNU Lesser
+// General Public License, version 3 or any later version (LGPLv3 or
+// later), or the GNU General Public License, version 2 (GPLv2), in all
+// cases as published by the Free Software Foundation.
+//
+
+package glusterfs
+
+import (
+	"bytes"
+	"encoding/gob"
+	"encoding/json"
+	"net/http"
+	"time"
+
+	"github.com/heketi/heketi/pkg/glusterfs/api"
+
+	"github.com/boltdb/bolt"
+	"github.com/lpabon/godbc"
+)
+
+var (
+	clusterHealth = "ClusterHealth"
+)
+
+type dbchecker struct {
+	CurrentTime time.Time
+}
+
+func (d *dbchecker) Marshal() ([]byte, error) {
+	var buffer bytes.Buffer
+	enc := gob.NewEncoder(&buffer)
+	err := enc.Encode(*d)
+
+	return buffer.Bytes(), err
+}
+
+func (d *dbchecker) Unmarshal(buffer []byte) error {
+	dec := gob.NewDecoder(bytes.NewReader(buffer))
+	return dec.Decode(d)
+}
+
+func (*dbchecker) BucketName() string {
+	return BOLTDB_BUCKET_DBHEALTHZ
+}
+
+func (d *dbchecker) Write(tx *bolt.Tx) error {
+	godbc.Require(tx != nil)
+	return EntrySave(tx, d, clusterHealth)
+}
+
+func (d *dbchecker) Read(tx *bolt.Tx) error {
+	godbc.Require(tx != nil)
+	return EntryLoad(tx, d, clusterHealth)
+}
+
+func checkDBState(db *bolt.DB, info *api.Healthz) {
+
+	var faulty = "faulty"
+	//db Read check
+	err := db.View(func(tx *bolt.Tx) error {
+		h := dbchecker{}
+		return h.Read(tx)
+	})
+
+	if err != nil && err != ErrNotFound {
+		info.ClusterHeath = faulty
+		info.DbState = faulty
+		return
+	}
+	info.DbState = "read"
+
+	//db write check
+	err = db.Update(func(tx *bolt.Tx) error {
+		h := dbchecker{CurrentTime: time.Now()}
+		return h.Write(tx)
+	})
+	if err == nil {
+		info.DbState = "write"
+	}
+	return
+}
+
+func checkClusterState(health *NodeHealthCache, info *api.Healthz) {
+
+	if len(health.nodes) == 0 {
+		return
+	}
+	totalUnhealthyNodes := 0
+	for _, n := range health.nodes {
+		if !n.Up {
+			totalUnhealthyNodes++
+		}
+		info.NodesStatus = append(info.NodesStatus, (api.NodeHealthStatus)(*n))
+	}
+	if totalUnhealthyNodes == len(info.NodesStatus) {
+		info.ClusterHeath = "faulty"
+	} else {
+		info.ClusterHeath = "healthy"
+	}
+	return
+}
+
+func (a *App) Healthz(w http.ResponseWriter, r *http.Request) {
+	info := api.Healthz{}
+	info.NodesStatus = make([]api.NodeHealthStatus, 0)
+
+	w.Header().Set("Content-Type", "application/json; charset=UTF-8")
+
+	// db status check
+	checkDBState(a.db, &info)
+
+	if !MonitorGlusterNodes {
+		if err := json.NewEncoder(w).Encode(info); err != nil {
+			panic(err)
+		}
+		return
+	}
+
+	// cluster status check
+	checkClusterState(a.nhealth, &info)
```

Comment:
I think this would read more logically to do the serialization once (at the bottom of the function) and move this line up into the if-staement (with an inverted condition) like so:
```golang
if MonitorGlusterNodes {
    checkClusterState(a.nhealth, &info)
}
// ... serialize response ...
```